text/0000-api-version-compat.md

```diff
@@ -0,0 +1,119 @@
+# Meta
+[meta]: #meta
+- Name: Lifecycle API version support changes
+- Start Date: 2020-04-14
+- Author(s): [Emily Casey](https://github.com/ekcasey)
+- RFC Pull Request: (leave blank)
+- CNB Pull Request: (leave blank)
+- CNB Issue: (leave blank)
+
+# Summary
+[summary]: #summary
+
+In order enable smoother migrations and upgrades this RFC proposes the following changes to the API version compatibility rules:
+
+1. When the lifecycle increments an API versions in the `0.x` version line, this shall interpreted as non-breaking
+1. The spec may make certain types of "breaking" changes in a `0.x` API version, so long as those changes do not preclude simultaneous lifecycle support for the previous minor or require collaborating buildpacks to implement the same API version.
+1. When the lifecycle implments a new `0.x` API minor version, it will treat the usage of features removed between minor spec versions as a deprecated.
+
+# Motivation
+[motivation]: #motivation
+
+## Non-Breaking 0.x API Versions
+Originally we specified that all minor version increments in the `0.x` line would be interpreted as breaking changes, when applying [compatibility rules](https://github.com/buildpacks/spec/blob/master/buildpack.md#buildpacktoml-toml).
+This was the correct decision initially b/c we needed a way to indicate breaking changes before the APIs reached 1.0.
+However, this has become annoying over time as it adds considerable friction to the release of non-breaking features. All platforms and buildpacks must explicitly update to support the new APIs before pulling in a new lifecycle, even when a new minor would not break them.
+As the community of platform and buildpack authors grows and matures, this pain is magnified.
+
+Now that the APIs have stabilized we can realistically guarantee that we will not make breaking lifecycle changes before the implementation of the 1.0 APIs are released.
+
+## Deprecations
+We will allow the spec to remove features in a minor API and interperet this as a deprecated feature in the lifecycle.
+Previous we never formalized rules around deprecations. This made sense because all API changes have been breaking.
+However, deprecations can aid platform and buldpack as they prepare to perform a major API upgrade.
```

Comment:
This is probably out of scope of this RFC, but there's a question of how we communicate deprecations to platform/buildpacks if it's end users that are seeing those notices through build output.

Reply:
Maybe `report.toml` :P, but I probably shouldn't treat that as a dumping ground.

Reply:
Do the deprecation modes address this concern? Platform and buildpack authors can turn on the deprecation output (or enable a failure mode as @jkutner suggested).